numba/targets/arraymath.py

```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+    if np.any(~np.isfinite(val)) or np.any(val < v_min) or np.any(val > v_max):
+        raise ValueError('unable to safely conform val to a.dtype')
+
+@register_jitable
+def _check_val_float(a, val):
+    finfo = np.finfo(a.dtype)
+    v_min = finfo.min
+    v_max = finfo.max
+
+    # check finite values are within bounds
+    finite_vals = val[np.isfinite(val)]
+    if np.any(finite_vals < v_min) or np.any(finite_vals > v_max):
+        raise ValueError('unable to safely conform val to a.dtype')
+
+@overload(np.fill_diagonal)
+def np_fill_diagonal(a, val, wrap=False):
+
+    def _abort_mission(*args):
+        raise ValueError("array must be at least 2-d")
+
+    def fill_diagonal_impl_scalar_val(a, val, wrap=False):
+        _fill_diagonal_scalar(a, val, wrap)
+
+    def fill_diagonal_impl_scalar_val_a_int(a, val, wrap=False):
+        _check_val_int(a, np.array([val]))
+        _fill_diagonal_scalar(a, val, wrap)
+
+    def fill_diagonal_impl_scalar_val_a_float(a, val, wrap=False):
+        _check_val_float(a, np.array([val]))
+        _fill_diagonal_scalar(a, val, wrap)
+
+    def fill_diagonal_impl_seq_val(a, val, wrap=False):
+        val = np.array(val).flatten()
+        _fill_diagonal(a, val, wrap)
+
+    def fill_diagonal_impl_seq_val_a_int(a, val, wrap=False):
+        val = np.array(val).flatten()
+        _check_val_int(a, val)
+        _fill_diagonal(a, val, wrap)
+
+    def fill_diagonal_impl_seq_val_a_float(a, val, wrap=False):
+        val = np.array(val).flatten()
+        _check_val_float(a, val)
+        _fill_diagonal(a, val, wrap)
+
+    def fill_diagonal_impl_array_val(a, val, wrap=False):
+        val = val.flatten()
+        _fill_diagonal(a, val, wrap)
+
+    def fill_diagonal_impl_array_val_check_a_int(a, val, wrap=False):
+        val = val.flatten()
+        _check_val_int(a, val)
+        _fill_diagonal(a, val, wrap)
+
+    def fill_diagonal_impl_array_val_check_a_float(a, val, wrap=False):
+        val = val.flatten()
+        _check_val_float(a, val)
+        _fill_diagonal(a, val, wrap)
+
+    if a.ndim < 2:
+        return _abort_mission
+
+    # the following can be simplified after #3088; until then, employ
+    # a basic mechanism for catching cases where val is of a type/value
+    # which cannot safely be cast to a.dtype
+
+    if isinstance(val, (types.Float, types.Integer, types.Boolean)):
+        if isinstance(a.dtype, types.Integer):
+            return fill_diagonal_impl_scalar_val_a_int
+        elif isinstance(a.dtype, types.Float):
+            return fill_diagonal_impl_scalar_val_a_float
+        else:
+            return fill_diagonal_impl_scalar_val
+
+    elif isinstance(val, (types.Tuple, types.Sequence)):
+        if isinstance(a.dtype, types.Integer):
+            return fill_diagonal_impl_seq_val_a_int
+        elif isinstance(a.dtype, types.Float):
+            return fill_diagonal_impl_seq_val_a_float
+        else:
+            return fill_diagonal_impl_seq_val
```

Comment:
Don't think the tests hit this line https://codecov.io/gh/numba/numba/compare/9fc637e3391e023b839afe00e6c86063c8cb82ad...c705734096b550f4ead5f050ca771ee884f8869f/src/numba/targets/arraymath.py#L1043 ?

Reply:
Added tests - this should now be covered - will keep an eye out